File: ruby/lib/closer/health.rb
# frozen_string_literal: true

require "webrick"
require "json"

module Closer
  # Lightweight HTTP health check server for container orchestration.
  class HealthServer
    def initialize(port:, logger: nil)
      @port = port
      @logger = logger
      @server = nil
    end

    def start
      @server = WEBrick::HTTPServer.new(
        Port: @port,
        Logger: WEBrick::Log.new("/dev/null"),
        AccessLog: []
      )

      @server.mount_proc "/healthz" do |_req, res|
        res.content_type = "application/json"
        res.body = '{"status":"healthy"}'
      end

      @server.mount_proc "/readyz" do |_req, res|
        res.content_type = "application/json"
        res.body = '{"status":"ready"}'
      end

      Thread.new { @server.start }
      @logger&.info("Health check server started on port #{@port}")
    end

    def stop
      @server&.shutdown
    end
  end
end

File: python/src/health.py
"""
HTTP health check endpoint for load balancers and container orchestrators.
Runs alongside the gRPC server on a separate port.
"""

import threading
import logging
from http.server import HTTPServer, BaseHTTPRequestHandler

logger = logging.getLogger(__name__)


class HealthHandler(BaseHTTPRequestHandler):
    service_name = "unknown"
    checks = {}

    def do_GET(self):
        if self.path == "/healthz":
            self.send_response(200)
            self.send_header("Content-Type", "application/json")
            self.end_headers()
            self.wfile.write(b'{"status":"healthy"}')
        elif self.path == "/readyz":
            # Readiness check - verify service dependencies
            self.send_response(200)
            self.send_header("Content-Type", "application/json")
            self.end_headers()
            self.wfile.write(b'{"status":"ready"}')
        else:
            self.send_response(404)
            self.end_headers()

    def log_message(self, format, *args):
        # Suppress default HTTP logging
        pass


def start_health_server(port: int, service_name: str = "unknown"):
    """Start HTTP health check server in a daemon thread."""
    HealthHandler.service_name = service_name
    server = HTTPServer(("0.0.0.0", port), HealthHandler)
    thread = threading.Thread(target=server.serve_forever, daemon=True)
    thread.start()
    logger.info(f"Health check server started on port {port}")
    return server
